### PR TITLE
Add support for BigQuery bulk export (to Avro, for now)

### PR DESCRIPTION
**Pull request by vinceatbluelabs**
This is just enough code to be able to move a table from BigQuery to BigQuery via a GCS bucket.  

It turns out that BigQuery doesn't export via Parquet, so the easiest path was to add the first steps of Avro support to Records Mover, and to allow Avro on BigQuery import.

In addition to that functionality, it lays the groundwork for general BigQuery export, which will be useful for moving to other places like Redshift as well.

To get BigQuery->Redshift to work for medium-sized tables, we'll probably need one of:

* Delimited support in BigQuery export (assuming Redshift and BigQuery share a delimited variant...)
* Add Avro import support to Redshift

To get BigQuery->Redshift to work for large tables, we'll need some kind of GCS->S3 copying at scale.  In typical vendor fashion, I think the GCP stuff I played around with in #132 is S3->GCS only.  We might also need to teach BigQuery, Redshift and the Records spec about compressed Avro files, as well.

### CODE REVIEW

File: records_mover/db/bigquery/unloader.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+import sqlalchemy
+import pprint
```
Comment (bluelabsbutler) [inline records_mover/db/bigquery/unloader.py:2]: F401 'pprint' imported but unused

File: tests/unit/db/bigquery/test_bigquery_unloader.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
+import unittest
+
+from records_mover.db.bigquery.unloader import BigQueryUnloader
+from records_mover.records.records_format import DelimitedRecordsFormat, ParquetRecordsFormat, AvroRecordsFormat
```
Comment (bluelabsbutler) [inline tests/unit/db/bigquery/test_bigquery_unloader.py:4]: F401 'records_mover.records.records_format.ParquetRecordsFormat' imported but unused

Comment (bluelabsbutler) [inline tests/unit/db/bigquery/test_bigquery_unloader.py:4]: E501 line too long (112 > 100 characters)

File: tests/unit/db/bigquery/test_bigquery_unloader.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+from records_mover.db.errors import NoTemporaryBucketConfiguration
+from mock import MagicMock, Mock
+from unittest.mock import patch, ANY
```
Comment (bluelabsbutler) [inline tests/unit/db/bigquery/test_bigquery_unloader.py:7]: F401 'unittest.mock.patch' imported but unused

File: tests/unit/records/sources/test_table_edge_cases.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+from records_mover.records.sources.table import TableRecordsSource
+from mock import Mock, patch, ANY
```
Comment (bluelabsbutler) [inline tests/unit/records/sources/test_table_edge_cases.py:2]: F401 'mock.ANY' imported but unused

File: tests/unit/records/sources/test_table_edge_cases.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import unittest
+from records_mover.records.targets.base import MightSupportMoveFromTempLocAfterFillingIt
```
Comment (bluelabsbutler) [inline tests/unit/records/sources/test_table_edge_cases.py:4]: F401 'records_mover.records.targets.base.MightSupportMoveFromTempLocAfterFillingIt' imported but unused

File: tests/unit/db/bigquery/test_bigquery_loader.py [hunk 6]
```diff
@@ -1,12 +1,19 @@
 import unittest
 
 from records_mover.db.bigquery.loader import BigQueryLoader
-from records_mover.records.records_format import DelimitedRecordsFormat, ParquetRecordsFormat
+from records_mover.records.records_format import (
```
Comment (bluelabsbutler) [inline tests/unit/db/bigquery/test_bigquery_loader.py:4]: This file, one of the 3 largest in the project, increased in size to 379 lines.  The total size of those files is now 3421 lines (target: 1103).  Is this file complex enough to refactor?

File: tests/unit/db/bigquery/test_bigquery_loader.py [hunk 7]
```diff
@@ -125,6 +132,33 @@ def test_can_load_this_format_true(self,
         mock_load_job_config.assert_called_with(set(), mock_load_plan)
         self.assertEqual(True, out)
 
+
+    @patch('records_mover.db.bigquery.loader.load_job_config')
```
Comment (bluelabsbutler) [inline tests/unit/db/bigquery/test_bigquery_loader.py:136]: E303 too many blank lines (2)

File: tests/unit/db/bigquery/test_bigquery_loader.py (hunk 6 above)
Comment (bluelabsbutler) [inline tests/unit/db/bigquery/test_bigquery_loader.py:4]: This file, one of the 3 largest in the project, increased in size to 378 lines.  The total size of those files is now 3420 lines (target: 1103).  Is this file complex enough to refactor?

File: metrics/bigfiles_high_water_mark [hunk 8]
```diff
@@ -1 +1 @@
-1103
+1125
```
Comment (vinceatbluelabs) [inline metrics/bigfiles_high_water_mark:1]:
```
389: tests/integration/*****
373: setup.py
364: records_mover/records/schema/field/__init__.py

Reduce total number of bigfiles violations to 1103 or below!
```

File: records_mover/db/bigquery/loader.py [hunk 9]
```diff
@@ -87,8 +87,6 @@ def load_from_fileobj(self, schema: str, table: str,
         # https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.load_table_from_file
         job = client.load_table_from_file(fileobj,
                                           f"{schema}.{table}",
-                                          # Must match the destination dataset location.
-                                          location="US",
```
Comment (vinceatbluelabs) [inline records_mover/db/bigquery/loader.py:91]: I'm not convinced this was ever needed - it's in the example code, but everything seems to run fine without it, so there must be some inference logic to figure out the dataset location based on the client object.

File: records_mover/db/bigquery/bigquery_db_driver.py [hunk 10]
```diff
@@ -110,6 +116,17 @@ def tweak_records_schema_for_load(self,
...
             return records_schema.convert_datetimes_to_datetimetz()
+        elif isinstance(records_format, AvroRecordsFormat):
+            # https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-avro#logical_types
+            #
+            # "Note: There is no logical type that directly
+            # corresponds to DATETIME, and BigQuery currently doesn't
+            # support any direct conversion from an Avro type into a
+            # DATETIME field."
+            #
+            return records_schema.convert_datetimes_to_string()
```
Comment (vinceatbluelabs) [inline records_mover/db/bigquery/bigquery_db_driver.py:130]: This came up during the table2table integration test; BigQuery fails to load if you try to load the same string it exports from a DATETIME column into a DATETIME column.

Comment (cwegrzyn) [reply records_mover/db/bigquery/bigquery_db_driver.py:130]: 😬

**Review by cwegrzyn — APPROVED** [review]
This looks great to me! Thanks for getting it in! :shipit:

**Pull request merged**